orchestrator/relayer/src/batch_relaying.rs

```diff
@@ -199,32 +205,44 @@ async fn submit_batches(
                 let total_cost = total_cost.unwrap();
                 let gas_price_as_f32 = downcast_to_f32(cost.gas_price).unwrap(); // if the total cost isn't greater, this isn't
 
-                info!(
-                    "We have detected latest batch {} but latest on Ethereum is {} This batch is estimated to cost {} Gas / {:.4} ETH to submit",
-                    latest_cosmos_batch_nonce,
-                    latest_ethereum_batch,
-                    cost.gas_price.clone(),
-                    total_cost / one_eth_f32()
-                );
-
-                cost.gas_price = ((gas_price_as_f32 * eth_gas_price_multiplier) as u128).into();
-
-                let res = send_eth_transaction_batch(
-                    current_valset.clone(),
-                    oldest_signed_batch,
-                    &oldest_signatures,
-                    timeout,
-                    gravity_contract_address,
-                    gravity_id.clone(),
-                    cost,
-                    eth_client.clone(),
+                if fee_manager.can_send_batch(
+                    &cost,
+                    &oldest_signed_batch.total_fee,
+                    &oldest_signed_batch.token_contract,
                 )
-                .await;
+                    .await
+                {
+                    let token_contract = oldest_signed_batch.token_contract;
+
+                    info!(
+                        "We have detected latest batch {} but latest on Ethereum is {} This batch is estimated to cost {} Gas / {:.4} ETH to submit",
+                        latest_cosmos_batch_nonce,
+                        latest_ethereum_batch,
+                        cost.gas_price.clone(),
+                        total_cost / one_eth_f32()
+                    );
 
-                if res.is_err() {
-                    warn!("Batch submission failed with {:?}", res);
+                    cost.gas_price = ((gas_price_as_f32 as u128 * eth_gas_price_multiplier as u128) as u128).into();
```

Comment:
could this overflow? The f32 to u128 cast looks iffy (and removes the decimal part if there is any?)

Reply:
yeah the original code use `((gas_price_as_f32 * eth_gas_price_multiplier) as u128).into();`

Reply:
It's safer to use `.into()` or `u128::from()` instead of `as` and `checked_mul` for overflow checking.

Reply:
I tried with `u128::from` but does not seem to support f32.

`into` also required a type?

Reply:
Compiler tries to infer type when you use `into`. But, if it isn't possible, you can use `from`.

Reply:
> I tried with `u128::from` but does not seem to support f32.

It implies that it may be unsafe to use. Anyway, you can revisit this later when [`FloatToInt`](https://doc.rust-lang.org/stable/std/convert/trait.FloatToInt.html) is stable.

Reply:
Read this for details on when this can be unsafe: https://doc.rust-lang.org/stable/std/primitive.f32.html#method.to_int_unchecked

Reply:
I see, I found it is possible to use `to_int_unchecked`

```  
                  cost.gas_price = ((gas_price_as_f32.to_int_unchecked::<u128>() * eth_gas_price_multiplier.to_int_unchecked::<u128>())
                        as u128)
                        .into();
  ```
  
  but I need to make the function unsafe and propagate the attribute all the way up.
  
  should I leave it as it is for now?

Reply:
Yes. That is fine for now.